### PR TITLE
Create secondary indices based on table bean annotations (#3923)

### PR DESCRIPTION
**Pull request by breader124**
Allow creation of secondary indices based on annotations put on beans

## Motivation and Context
This changes solves the problem described in the #3923 

## Modifications
I've implemented the mechanism that inside the `createTable()` method of the `DefaultDynamoDbTable` class takes the `tableSchema` field and uses it to come up with the list of LSIs and GSIs in the table's schema. They are distinguished based on the fact that primary partition key of the table should be the same attribute as the partition key of the LSI. If that's the true, then index is being mapped to LSI. Otherwise, it's treated as GSI and and will be mapped accordingly. 

During the mapping there are no information about projections and provisioned throughput available (only applicable to GSIs). I believe that it should be the scope of another bunch of work to make it possible on the bean level to specify projections and desired provisioned throughput and that's why I tried to come up with some sensible defaults for now:
- for projections I used `ProjectionType.ALL`, it may incur some additional cost for people that wouldn't need it, but after all it'll allow to use the index for all imaginable cases
- for provisioned throughput I used the defaults of 20 RCUs and 20 WCUs

## Testing
I've implemented additional tests in `DefaultDynamoDbTableTest` class to make sure that bugfix actually works. All tests executed while running `./mvnw package` passed.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by breader124** [comment]
I see you've been assigned as a reviewer @dagnir, would you find some time to give this PR a look, please?

**Comment by hamburml** [comment]
I need this :)

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/client/DefaultDynamoDbTable.java [hunk 1]
```diff
@@ -115,7 +125,52 @@ public void createTable(Consumer<CreateTableEnhancedRequest.Builder> requestCons
...
+        return indicesGroups.getOrDefault(IndexType.GSI, emptyList()).stream()
+                            .map(this::mapIndexMetadataToEnhancedGlobalSecondaryIndex)
+                            .collect(Collectors.toList());
+    }
+
+    private EnhancedGlobalSecondaryIndex mapIndexMetadataToEnhancedGlobalSecondaryIndex(IndexMetadata indexMetadata) {
+        return EnhancedGlobalSecondaryIndex.builder()
+                                           .indexName(indexMetadata.name())
+                                           .projection(pb -> pb.projectionType(ProjectionType.ALL))
+                                           .provisionedThroughput(ptb -> ptb.readCapacityUnits(20L).writeCapacityUnits(20L))
```
[the remark this answers is not in the document]
Comment (breader124) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/client/DefaultDynamoDbTable.java]: I was hesitant as well. I was thinking that even when the billing mode for the table is `PAY_PER_REQUEST`, there's need to specify provisioned throughput for GSIs. It seems you are right though and there's no need to do so. Fixed

[the remark this answers is not in the document]
Comment (L-Applin) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/client/DefaultDynamoDbTable.java]: I ran a quick test on my end, creating a table using the code from this revision (without the provisioned throughput). The test created the table with the `On-Demand` billing mode. I think that's what we want. Thanks!

**Comment by L-Applin** [comment]
It seems that if the TableSchema is created statically using `documentSchemaBuilder`, when no secondary indexes are specified, the create-request is made with empty secondary index list

```
        docMappedtable = enhancedClient.table(tableName,
                                              TableSchema.documentSchemaBuilder()
                                                         .addIndexPartitionKey(TableMetadata.primaryIndexName(),
                                                                               "id",
                                                                               AttributeValueType.S)
                                                         .addIndexSortKey(TableMetadata.primaryIndexName(), "sort", AttributeValueType.S)
                                                         .attributeConverterProviders(defaultProvider())
                                                         .build());
        docMappedtable.createTable();
```

**Comment by breader124** [comment]
I just fixed it. All unit tests for the `services-custom/dynamodb-enhanced` submodule are not passing

**Comment by L-Applin** [comment]
> I just fixed it. All unit tests for the `services-custom/dynamodb-enhanced` submodule are now passing

Awesome, I'm re-running test now!

**Comment by L-Applin** [comment]
all checks passes, approved!

**Review by L-Applin — APPROVED** [review]

**Comment by debora-ito** [comment]
@all-contributors please add @breader124 for code.

**Pull request merged**